1-Draft/RFCNNNN-Propagate-Execution-Preferences-Beyond-Module-Scope.md

```diff
@@ -0,0 +1,133 @@
+I should be able to change a published command from an advanced function to a cmdlet or vice versa,
+so that as long as I keep the same functionality in the command, scripts and modules where that command is in use should behave the same way.
+
+## User experience
+
+```powershell
+# First, create a folder for a new module
+$moduleName = 'RFCPropagateExecPrefDemo'
+$modulePath = Join-Path -Path $([Environment]::GetFolderPath('MyDocuments')) -ChildPath PowerShell/Modules/${moduleName}
+New-Item -Path $modulePath -ItemType Directory -Force > $null
+
+# Then, create the manifest (which would have the PersistCommandExecutionPreferences
+# optional feature enabled by default, to correct the behaviour moving forward in a
+# non-breaking way; downlevel versions of PowerShell would ignore the optional feature
+# flags)
+$nmmParameters = @{
+    Path = "${modulePath}/${moduleName}.psd1"
+    RootModule = "./${moduleName}.psm1"
+    FunctionsToExport = @('Test-1')
+    PassThru = $true
+}
+New-ModuleManifest @nmmParameters | Get-Content
+
+# Output:
+#
+# @{
+#
+# RootModule = './RFCPropagateExecPrefDemo.psm1'
+#
+# # Private data to pass to the module specified in RootModule/ModuleToProcess. This may also contain a PSData hashtable with additional module metadata used by PowerShell.
+# PrivateData = @{
+#
+#     <snip>
+#
+#     PSData = @{
+#
+#         # Optional features enabled in this module.
+#         OptionalFeatures = @(
+#             'PersistCommandExecutionPreferences'
+#         )
+#
+#         <snip>
+#
+#     } # End of PSData hashtable
+#
+#     <snip>
+#
+# } # End of PrivateData hashtable
+#
+# }
+
+# Then, create the script module file, along with a second module in memory that it invokes,
+# and import both modules
+$scriptModulePath = Join-Path -Path $modulePath -ChildPath ${moduleName}.psm1
+New-Item -Path $scriptModulePath -ItemType File | Set-Content -Encoding UTF8 -Value @'
+    function Test-1 {
+        [cmdletBinding()]
+        param()
+        Test-2
+    }
+'@
+Import-Module $moduleName
+New-Module Name test2 {
+    function Test-2 {
+        [cmdletBinding()]
+        param()
+        Write-Verbose 'Verbose output'
+    }
+} | Import-Module
+
+# When invoking the Test-2 command with -Verbose, it shows verbose output, as expected
+Test-2 -Verbose
+
+# Output:
+#
+# VERBOSE: Verbose output
+
+# Thanks to this feature, when invoking the Test-1 command with -Verbose, it also shows
+# verbose output. In PowerShell 6.2 and earlier, no verbose output would appear as a
+# result of this command due to the issue preventing execution preferences from propagating
+# beyond module/script scope
+Test-1 -Verbose
+
+# Output:
+#
+# VERBOSE: Verbose output
+```
+
+## Specification
+
+To resolve this problem, a new optional feature called `PersistCommandExecutionPreferences` would be defined in PowerShell. When this feature is enabled in a script or module, it would change how common parameters work in that script or module.
```

Comment:
I am wondering if it is better to rather have an optional feature for the contrary, i.e. a feature flag for legacy behaviour. I fear that adoption and feedback will not be good if the RFC changes are not on by default. Yes, I know, the team does not want to make PS5->7 migration harder but it feels right to fix fundamental behaviour like this.

Reply:
You know, for that RFC, that's a great point. It would fix many problems, and likely introduce some that need  fixing, but by having it on by default with the option to disable it, we would be guiding users in the right direction: fix your scripts now or disable this and fix them later when you have time. I like it, will update the RFC with the suggestion in the alternate proposals section when back in my office. Thanks!

Reply:
You know, for that RFC, that's a great point. It would fix many problems, and likely introduce some that need  fixing, but by having it on by default with the option to disable it, we would be guiding users in the right direction: fix your scripts now or disable this and fix them later when you have time. I like it, will update the RFC with the suggestion in the alternate proposals section when back in my office. Thanks!

Reply:
Note to self: this highlights the need to be able to have optional features on at one level and for scripters to be able to turn them off at another. Just adding this note so that I add an additional proposal to the optional feature RFC as well later.

Reply:
PowerShell recently added support for experimental features, maybe just using a different word other than experimental would make thatsuitable for a feature flag for fallback to legacy behaviour

Reply:
I updated this last night. I think experimental feature flags serve a different purpose. When you have time, let me know what you think about the added alternate proposals.

Reply:
Looks good to me :)